pkg/genericapiserver/serialization_config.go

```
@@ -0,0 +1,106 @@
+	InternalResourceEncodings map[string]unversioned.GroupVersion
+}
+
+func NewDefaultResourceEncodingConfig() *DefaultResourceEncodingConfig {
+	return &DefaultResourceEncodingConfig{Groups: map[string]*GroupResourceEncodingConfig{}}
+}
+
+func newGroupResourceEncodingConfig(defaultEncoding, defaultInternalVersion unversioned.GroupVersion) *GroupResourceEncodingConfig {
+	return &GroupResourceEncodingConfig{
+		DefaultExternalEncoding: defaultEncoding, ExternalResourceEncodings: map[string]unversioned.GroupVersion{},
+		DefaultInternalEncoding: defaultInternalVersion, InternalResourceEncodings: map[string]unversioned.GroupVersion{},
+	}
+}
+
+func (o *DefaultResourceEncodingConfig) SetVersionEncoding(group string, externalEncodingVersion, internalVersion unversioned.GroupVersion) {
+	_, groupExists := o.Groups[group]
+	if !groupExists {
+		o.Groups[group] = newGroupResourceEncodingConfig(externalEncodingVersion, internalVersion)
+	}
+
+	o.Groups[group].DefaultExternalEncoding = externalEncodingVersion
+	o.Groups[group].DefaultInternalEncoding = internalVersion
+}
+
+func (o *DefaultResourceEncodingConfig) SetResourceEncoding(resourceBeingStored unversioned.GroupResource, externalEncodingVersion, internalVersion unversioned.GroupVersion) {
+	group := resourceBeingStored.Group
+	_, groupExists := o.Groups[group]
+	if !groupExists {
+		o.Groups[group] = newGroupResourceEncodingConfig(externalEncodingVersion, internalVersion)
+	}
+
+	o.Groups[group].ExternalResourceEncodings[resourceBeingStored.Resource] = externalEncodingVersion
+	o.Groups[group].InternalResourceEncodings[resourceBeingStored.Resource] = internalVersion
+}
+
+func (o *DefaultResourceEncodingConfig) StoragageEncodingFor(resource unversioned.GroupResource) unversioned.GroupVersion {
+	groupEncoding, groupExists := o.Groups[resource.Group]
+	if !groupExists {
+		// return the most preferred external version for the group
+		return registered.GroupOrDie(resource.Group).GroupVersion
```

Its probably better to return an error than die?